File: provider/src/main/java/au/com/dius/pact/provider/junitsupport/loader/PactUrlLoader.java
package au.com.dius.pact.provider.junitsupport.loader;

import au.com.dius.pact.core.model.DefaultPactReader;
import au.com.dius.pact.core.model.Pact;
import au.com.dius.pact.core.model.UrlsSource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.toList;

/**
 * Implementation of {@link PactLoader} that downloads pacts from given urls
 */
public class PactUrlLoader implements PactLoader {
    private static final Logger LOGGER = LoggerFactory.getLogger(PactUrlLoader.class);
    private final String[] urls;
    private final UrlsSource pactSource;

  public PactUrlLoader(final String[] urls) {
    this.urls = urls;
    this.pactSource = new UrlsSource(Arrays.stream(urls).collect(Collectors.toList()));
  }

  public PactUrlLoader(final PactUrl pactUrl) {
        this(pactUrl.urls());
    }

  @Override
  public String description() {
    return "URL(" + Arrays.toString(urls) + ")";
  }

  public List<Pact> load(final String providerName) {
      return Arrays.stream(urls)
        .map(url -> {
          Pact pact = DefaultPactReader.INSTANCE.loadPact(url);
          this.getPactSource().getPacts().put(url, pact);
          return pact;
        })
        .collect(toList());
  }

  @Override
  public UrlsSource getPactSource() {
    return pactSource;
  }
}

File: provider/src/main/kotlin/au/com/dius/pact/provider/junitsupport/loader/PactUrlLoader.kt
package au.com.dius.pact.provider.junitsupport.loader

import au.com.dius.pact.core.model.DefaultPactReader
import au.com.dius.pact.core.model.Pact
import au.com.dius.pact.core.model.PactReader
import au.com.dius.pact.core.model.PactSource
import au.com.dius.pact.core.model.UrlSource
import au.com.dius.pact.core.model.UrlsSource
import au.com.dius.pact.core.support.Auth
import au.com.dius.pact.core.support.expressions.SystemPropertyResolver
import au.com.dius.pact.core.support.expressions.ValueResolver

/**
 * Implementation of [PactLoader] that downloads pacts from given urls
 */
open class PactUrlLoader(val urls: Array<String>, val authentication: Auth? = null) : PactLoader {
  lateinit var pactSource: UrlsSource
  var pactReader: PactReader = DefaultPactReader
  var resolver: ValueResolver = SystemPropertyResolver

  constructor(pactUrl: PactUrl) : this(pactUrl.urls, when {
    pactUrl.auth.token.isNotEmpty() -> Auth.BearerAuthentication(pactUrl.auth.token)
    pactUrl.auth.username.isNotEmpty() -> Auth.BasicAuthentication(pactUrl.auth.username,
      pactUrl.auth.password)
    else -> null
  })

  override fun description() = "URL(${urls.contentToString()})"

  override fun load(providerName: String): List<Pact> {
    pactSource = UrlsSource(urls.asList())
    return urls.map { url ->
      val options = mutableMapOf<String, Any>()
      if (authentication != null) {
        options["authentication"] = authentication.resolveProperties(resolver)
      }
      val pact = pactReader.loadPact(UrlSource(url), options)
      pactSource.addPact(url, pact as Pact)
      pact
    }
  }

  override fun getPactSource(): PactSource {
    return pactSource
  }
}
